firedrake/preconditioners/pmg.py

```diff
@@ -0,0 +1,290 @@
+        fV = fctx.J.arguments()[0].function_space()
+
+        cbcs = cctx._problem.bcs
+        fbcs = fctx._problem.bcs
+
+        I = prolongation_matrix(fV, cV, fbcs, cbcs)
+        R = PETSc.Mat().createTranspose(I)
+        return R, None
+
+    def view(self, pc, viewer=None):
+        if viewer is None:
+            viewer = PETSc.Viewer.STDOUT
+        viewer.printfASCII("p-multigrid PC\n")
+        self.ppc.view(viewer)
+
+
+def prolongation_transfer_kernel(Pk, P1):
+    # Works for Pk, Pm; I just retain the notation
+    # P1 to remind you that P1 is of lower degree
+    # than Pk
+    from tsfc import compile_expression_dual_evaluation
+    from tsfc.fiatinterface import create_element
+    from firedrake import TestFunction
+
+    expr = TestFunction(P1)
+    coords = Pk.ufl_domain().coordinates
+    to_element = create_element(Pk.ufl_element(), vector_is_mixed=False)
+
+    ast, oriented, needs_cell_sizes, coefficients, _ = compile_expression_dual_evaluation(expr, to_element, coords, coffee=False)
+    kernel = op2.Kernel(ast, ast.name)
+    return kernel
+
+
+def prolongation_matrix(Pk, P1, Pk_bcs, P1_bcs):
+    sp = op2.Sparsity((Pk.dof_dset,
+                       P1.dof_dset),
+                      (Pk.cell_node_map(),
+                       P1.cell_node_map()))
+    mat = op2.Mat(sp, PETSc.ScalarType)
+    mesh = Pk.ufl_domain()
+
+    fele = Pk.ufl_element()
+    if isinstance(fele, MixedElement) and not isinstance(fele, (VectorElement, TensorElement)):
+        for i in range(fele.num_sub_elements()):
+            Pk_bcs_i = [bc for bc in Pk_bcs if bc.function_space().index == i]
+            P1_bcs_i = [bc for bc in P1_bcs if bc.function_space().index == i]
+
+            rlgmap, clgmap = mat[i, i].local_to_global_maps
+            rlgmap = Pk.sub(i).local_to_global_map(Pk_bcs_i, lgmap=rlgmap)
+            clgmap = P1.sub(i).local_to_global_map(P1_bcs_i, lgmap=clgmap)
+            unroll = any(bc.function_space().component is not None
+                         for bc in chain(Pk_bcs_i, P1_bcs_i) if bc is not None)
+            matarg = mat[i, i](op2.WRITE, (Pk.sub(i).cell_node_map(), P1.sub(i).cell_node_map()),
+                               lgmaps=(rlgmap, clgmap), unroll_map=unroll)
+            op2.par_loop(prolongation_transfer_kernel(Pk.sub(i), P1.sub(i)), mesh.cell_set,
+                         matarg)
+
+    else:
+        rlgmap, clgmap = mat.local_to_global_maps
+        rlgmap = Pk.local_to_global_map(Pk_bcs, lgmap=rlgmap)
+        clgmap = P1.local_to_global_map(P1_bcs, lgmap=clgmap)
+        unroll = any(bc.function_space().component is not None
+                     for bc in chain(Pk_bcs, P1_bcs) if bc is not None)
+        matarg = mat(op2.WRITE, (Pk.cell_node_map(), P1.cell_node_map()),
+                     lgmaps=(rlgmap, clgmap), unroll_map=unroll)
+        op2.par_loop(prolongation_transfer_kernel(Pk, P1), mesh.cell_set,
+                     matarg)
```

Comment:
This suggests we should provide a better API for doing this kind of thing. But not here.

Reply:
It feels like `make_interpolator` from `interpolation.py` almost does this.

Reply:
Probably needs too much refactoring.